rfd/0202-db-multi-session-mfa.md

```diff
@@ -0,0 +1,270 @@
+            tsh ->> Teleport: GenerateUserCerts with cached MFA response
+            Teleport ->> tsh: User cert with database route
+        end
+
+        tsh <<->> Teleport: database connection with TLS routing
+        tsh ->> user: output
+    end
+```
+
+In favor of better UX, no new role option is introduced for multi-session MFA.
+The multi-session MFA extends [RFD 155 Scoped Webauthn
+Credentials](https://github.com/gravitational/teleport/blob/master/rfd/0155-scoped-webauthn-credentials.md)
+by allowing reuse for `ChallengeScope_CHALLENGE_SCOPE_USER_SESSION`.
+
+Similar to `SCOPE_ADMIN_ACTION`, `ChallengeScope_CHALLENGE_SCOPE_USER_SESSION`
+now can be requested with `CHALLENGE_ALLOW_REUSE_YES`. The responded session data
+will be reusable until it expires. Clients must go through MFA ceremony again if
+it expires. Currently, reuse of the session data is allowed within 5 minutes.
+This duration is hard-coded and controlled on the server side. We could
+introduce a new option to override the period but leaving it out of scope for
+this RFD for now.
+
+The reusable MFA response will be checked upon auth call of `GenerateUserCerts`
+where user requests a TLS user cert with database route. New logic is added to
+`GenerateUserCerts` when the request is identified from `tsh db exec` sessions:
+
+```diff
+// Requester is the name of the service that sent the request.
+ enum Requester {
+...
+    // TSH_APP_LOCAL_PROXY is set when the request was sent by a tsh app local proxy.
+    TSH_APP_LOCAL_PROXY = 4;
++   // TSH_DB_EXEC is set when the request was sent by "tsh db exec".
++   TSH_DB_EXEC = 5;
+}
+```
+
+And here goes the logic for handling `TSH_DB_EXEC`:
+- The purpose must be single use and TTL is constrained to single use TTL of one
+  minute.
+- The request must be for a database route and not for any other resources.
+- `CHALLENGE_ALLOW_REUSE_YES` when validating the MFA response.
+
+Here is a matrix of different modes that `tsh` can request for user certs when
+session MFA is required:
+
+| Command          | Reusable  | Cert TTL                                                        |
+|------------------|-----------|-----------------------------------------------------------------|
+| `tsh db connect` | no        | 1 minute                                                        |
+| `tsh proxy db`   | no        | min(identity.expires, `role.options.mfa_verification_interval`) |
+| `tsh db exec`    | 5 minutes | 1 minute                                                        |
+
+
+### The `tsh db exec` command
+
+General flow of the command:
+- Create auth client which should be reused for all API calls.
+- Fetch databases (either specified directly or through search).
+- For each database:
+  - Check MFA requirement:
+    - No MFA prompt if not required.
+    - MFA prompt if `per_session` MFA is required.
+    - MFA prompt if first time requesting MFA for `multi_session` or cached
```

Comment:
The client doesn't actually know when the MFA response expires - it's 5m on the backend. Will the client retry when a GenerateUserCerts request fails with `Webauthn session data not found`? Or will the client just estimate the timeout as 5 minutes from when the CreateAuthenticateChallenge request was sent?

Reply:
guessing the timeout on client is definitely not reliable if we can the backend logic one day.

i was thinking to check against a specific error when calling `GenerateUserCerts`, and i can add this new error message when failed to `ValidateMFAAuthResponse` with the reusable user session, maybe something like `failed to validate reusable MFA response ...`. but i don't know if there is a reliable way to detect if session is expired (different errors for different backends?).

maybe we can add a hint on the expiration in `MFAAuthenticateResponse`?

Reply:
All backends should just return a not found error. So you could return a new error in `ValidateMFAAuthResponse`, just make it return when you get a not found error and `requiredExtensions` = `{Scope: DB_EXEC, REUSE: true}`.

Reply:
Thanks. The backend and error details are implementation details IMO but i've added some text stating the expired session is detected by expecting specific errors.  2db1a19

> {Scope: DB_EXEC, REUSE: true}

In my previous update, i've dropped the new scope in favor of a new requester name. Do you think the new requester name is sufficient or a new scope is preferred??